libraries/AP_MSP/msp.cpp

```diff
@@ -0,0 +1,296 @@
+#include <AP_Math/AP_Math.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include "msp.h"
+#include <stdint.h>
+
+#if  HAL_MSP_ENABLED
+
+uint8_t MSP::msp_serial_checksum_buf(uint8_t checksum, const uint8_t *data, int len)
+{
+    while (len-- > 0) {
+        checksum ^= *data++;
+    }
+    return checksum;
+}
+
+int MSP::msp_serial_send_frame(msp_port_t *msp, const uint8_t * hdr, int hdr_len, const uint8_t * data, int data_len, const uint8_t * crc, int crc_len)
+{
+    if (msp->uart == nullptr) {
+        return 0;
+    }
+
+    // We are allowed to send out the response if
+    //  a) TX buffer is completely empty (we are talking to well-behaving party that follows request-response scheduling;
+    //     this allows us to transmit jumbo frames bigger than TX buffer (serialWriteBuf will block, but for jumbo frames we don't care)
+    //  b) Response fits into TX buffer
+    const uint32_t total_frame_length = hdr_len + data_len + crc_len;
+
+    if (!msp->uart->tx_pending() && (msp->uart->txspace() < total_frame_length)) {
+        return 0;
+    }
+
+    // Transmit frame
+    msp->uart->write(hdr, hdr_len);
+    msp->uart->write(data, data_len);
+    msp->uart->write(crc, crc_len);
+
+    return total_frame_length;
+}
+
+int MSP::msp_serial_encode(msp_port_t *msp, msp_packet_t *packet, msp_version_e msp_version)
+{
+    static const uint8_t msp_magic[MSP_VERSION_COUNT] = MSP_VERSION_MAGIC_INITIALIZER;
+    const int data_len = sbuf_bytes_remaining(&packet->buf);
```

Comment:
If this is ever negative there's been a few issues.

Also, `data_len` seems a strange name for space-remaining.

Reply:
yep, added a comment to explain why